exchanges/bitfinex/bitfinex_websocket.go

```diff
@@ -1663,35 +1646,139 @@ func (b *Bitfinex) ConfigureWS() error {
 	})
 }
 
-// Unsubscribe sends a websocket message to stop receiving data from the channel
-func (b *Bitfinex) Unsubscribe(channelsToUnsubscribe []stream.ChannelSubscription) error {
+// Subscribe sends a websocket message to receive data from channels
+func (b *Bitfinex) Subscribe(channels []stream.ChannelSubscription) error {
+	return b.parallelChanOp(channels, b.subscribeToChan)
+}
+
+// Unsubscribe sends a websocket message to stop receiving data from channels
+func (b *Bitfinex) Unsubscribe(channels []stream.ChannelSubscription) error {
+	return b.parallelChanOp(channels, b.unsubscribeFromChan)
+}
+
+// parallelChanOp performs a single method call in parallel across streams and waits to return any errors
+func (b *Bitfinex) parallelChanOp(channels []stream.ChannelSubscription, m func(*stream.ChannelSubscription) error) error {
+	wg := sync.WaitGroup{}
+	wg.Add(len(channels))
+	errC := make(chan error, len(channels))
+
+	for i := range channels {
+		go func(c *stream.ChannelSubscription) {
+			defer wg.Done()
+			if err := m(c); err != nil {
+				errC <- err
+			}
+		}(&channels[i])
+	}
+
+	wg.Wait()
+	close(errC)
+
 	var errs error
-	for i := range channelsToUnsubscribe {
-		idAny, ok := channelsToUnsubscribe[i].Params["chanId"]
-		if !ok {
-			errs = common.AppendError(errs, fmt.Errorf("cannot unsubscribe from a channel without an id"))
-			continue
-		}
-		chanID, ok := idAny.(int)
-		if !ok {
-			errs = common.AppendError(errs, fmt.Errorf("chanId is not an int"))
-			continue
-		}
+	for err := range errC {
+		errs = common.AppendError(errs, err)
+	}
 
-		req := map[string]interface{}{
-			"event":  "unsubscribe",
-			"chanId": chanID,
-		}
+	return errs
+}
 
-		err := b.Websocket.Conn.SendJSONMessage(req)
-		if err != nil {
-			errs = common.AppendError(errs, err)
-			continue
+// subscribeToChan handles a single subscription and parses the result
+// on success it adds the subscription to the websocket
+func (b *Bitfinex) subscribeToChan(c *stream.ChannelSubscription) error {
+	req := make(map[string]interface{})
+	req["event"] = "subscribe"
+	req["channel"] = c.Channel
+
+	for k, v := range c.Params {
+		// Resubscribing channels might already have this set
+		if k != "chanId" {
+			req[k] = v
 		}
-		// We do this before the unsubscribed event comes back so we can subscribe again when called from ResubcribeToChannel
-		b.Websocket.RemoveSuccessfulUnsubscriptions(channelsToUnsubscribe[i])
 	}
-	return errs
+
+	// Although docs only mention this for wsBook, it works for all chans
+	subID := strconv.FormatInt(b.Websocket.Conn.GenerateMessageID(false), 10)
```

Comment:
I think the only thing I will posit here is that we know the chanID in advance and it takes us all the way to receiving the response before we consider it a valid ID. So with the whole dropping of data on ln:154, we should technically know if the chanID is something that we're waiting on. Like, we could fall back on this number in the event of `b.Websocket.GetSubscription(chanID)` failure because the acknowledgement has not been processed in the time it takes for an update to come through.

I understand that this adds additional complexity, so I will not require it, but thought I'd float an additional possibility to resolve the dropped-data potential

Reply:
To further this point, I believe its the cause for the WS errors I was receiving:
```
[DEBUG]  | WEBSOCKET | 24/10/2023 13:10:43 | Bitfinex websocket connection: message received: {"event":"subscribed","channel":"book","chanId":266872,"symbol":"tBTCUSD","prec":"R0","freq":"F0","len":"100","subId":"155672758","pair":"BTCUSD"}
```
<img width="2365" alt="image" src="https://github.com/thrasher-corp/gocryptotrader/assets/9261323/0091bc72-6b03-4423-b4f7-43e37d5a681a">

<img width="1555" alt="image" src="https://github.com/thrasher-corp/gocryptotrader/assets/9261323/18e740d9-ceb4-41f2-91c5-01b32eb2dfa5">

Reply:
Yeah, I had exactly the same issue, bit didn't want to scope creep this first PR.
Pretty sure This is completely fixed in #1358 already, specifically f40636d9,
However I don't think I also addressed bitfinex's version of this problem.
Let me consider backmerging that.

Reply:
Nope; this does not fly.

We send in subID but we _only_ get it back in the subscribed response OR error on subscribe.
And the subID doesn't replace the chanID at all. It literally just gets repeated back once, during the subscribe response.
We don't get it back for any channel messages "after" that.

That means we couldn't tie any channel messages to the subscription by using it.
Until we get back the chanID for the sub, we won't have a viable key to use.

*However* I think just use the DefaultKey mechanism.
Two choices:
* We key first with DefaultKey, and re-add with ChanID after, and have a fallback when looking
* We just use DefaultKey throughout, and find the ChanId again just for unsubscribe.
This has the added benefit that people can unsubscribe without knowing ChanID.

I'm going for solution 2.

Reply:
Duh. None of that works.
We don't *have* anything to key off yet because the default subscription uses Pair, etc.

All we have for each update is literally the ChanID. That's it.

The problem is going down the InocmingData channel introduces a switch.
The other problem is that's where we have the real ChannelSubscription with the right data.
So the best thing to do has to be to AddSubscription before going down the IncomingData channel, but WITH the full channel.
I can do that. On it now. Again. Last time. Promise. Maybe.

Reply:
Fixed 29954a8